### ISSUE TITLE
Prevent change shell during setup

### ISSUE THREAD

**Issue by flip111** [robbyrussell/oh-my-zsh#4261]
Is there a flag to the setup script that prevents changing to zsh after installation? I would like zsh to become the default shell but not on the shell session during installation. This is because i have an installation script which halts when installing oh my zsh. I have to manually type exit before the script resumes in the old shell.

**Comment by apjanke** [comment]
There isn't. But this might be nice to have. (And if you want to go fully-scriptable, you'd probably want is a full "non-interactive" flag for the installer, because there are other points it might block and wait on user input as well, like the password for `chsh`.) The line at the end where it sources `~/.zshrc` after returning from the subshell (or inside the subshell, depending on how the installer was run) may be an issue too.

Since the installer is a core part of OMZ, changes to it take a while to go through. However, the installer doesn't care where it's run from, so you can easily make your own variant of it. Just create a new repo, copy the installer over, remove those last couple lines, and use the URL to your installer instead of the default one when doing your scripted installations.

**Comment by mcornella** [comment]
> Just create a new repo, copy the installer over, remove those last couple
lines, and use the URL to your installer instead of the default one when
doing your scripted installations.

That's how babun does it anyway, I think. A simple git clone and the zshrc
file modification is all that's needed, in fact.

**Comment by flip111** [comment]
I don't find the trigger line for starting the actual z-shell. There are a lot of things being sourced though

**Comment by apjanke** [comment]
I'm not sure I understand what you're saying here.

Aside from its installer, nothing in Oh My Zsh starts `zsh`. It works the other way around: you start `zsh` somehow (for example, by starting a new terminal session), and `zsh` automatically loads `~/.zshrc` as part of its startup sequence, and then `~/.zshrc` loads Oh My Zsh by sourcing `~/.oh-my-zsh/oh-my-zsh.sh`.

**Comment by flip111** [comment]
@apjanke i observed that after running `sh -c "$(curl -fsSL https://raw.github.com/robbyrussell/oh-my-zsh/master/tools/install.sh)"` the prompt changes, then when i type `exit` i return to me previous prompt and my setup script continues. This is made me to belief that zsh was started as sub-shell by oh-my-zsh.

**Comment by apjanke** [comment]
Ah: yes, during the installer, it is. But that's just a one-time quirk of the installer (sort of giving you a preview of what your `zsh` will look like from now on), and doesn't reflect normal usage of OMZ. (There's a longish discussion of this in #3664.)

Once it's installed, the normal sequence is: you start up a shell session, and OMZ is loaded automatically from `~/.zshrc`.

**Comment by flip111** [comment]
Is it possible to disable this functionality? (the one-time quirk of the installer)

**Comment by apjanke** [comment]
Easy change. Just comment out [these last two lines](https://github.com/robbyrussell/oh-my-zsh/blob/ccaaa992c733e4e3a4904d78ff84748c56236860/tools/install.sh#L60-L61)

```
env zsh
. ~/.zshrc
```

**Comment by flip111** [comment]
thanks @apjanke shall i leave the issue open for the official option?

**Comment by mcornella** [comment]
We can open a tracking issue related to the install script and point here. There are a number of issues with it and we should collect them. It doesn't matter if this one is closed, we got the feature request and we'll put it there.
**TL;DR** you can close this one thanks :grin:

**Comment by leoj3n** [comment]
Could wrap the code in a file descriptor check:

```
# Do not prompt user unless STDIN file descriptor is bound to TTY,
if [ -t 0 ]; then
  # chsh
  # env zsh
fi
```

Then run the install like:

```
0>/dev/null sh -c "$(curl -fsSL https://raw.githubusercontent.com/robbyrussell/oh-my-zsh/master/tools/install.sh)"
```

**Comment by lezsakdomi** [comment]
I've done some hacking, for automating OMZ installation process:

```bash
sh -c "$(curl -fsSL https://raw.github.com/robbyrussell/oh-my-zsh/master/tools/install.sh | sed '/\s*env\s\s*zsh\s*/d')" \
```

For other modifications, it's better to revert them after running the install script. See http://lezsakdomi-home.herokuapp.com/.bashrc and grep for `oh-my-zsh`

**Comment by Tomotoes** [comment]
```sh
NO_INTERACTIVE=true sh -c "$(curl -fsSL https://raw.githubusercontent.com/subtlepseudonym/oh-my-zsh/feature/install-noninteractive/tools/install.sh)"

```

**Comment by mcornella** [comment]
Hi there, I pulled the changes from #7266 in #5169, which mean you can now add `--skip-chsh` to the install script call, as well as set the environment variable `CHSH` to `no` so that the script doesn't attempt to change the default shell. You can also run it with `--unattended` so that zsh is not run at the end of the install. 

If you want to help speed up the testing and merge (which should happen early next week), you can wget or curl it from the URL `https://raw.githubusercontent.com/mcornella/oh-my-zsh/refactor-installer/tools/install.sh`. Read the comments at the top of the script to know how to call it.

Thanks!
